lib/XML/LibXML/Node.pm

```diff
@@ -108,25 +112,422 @@ role XML::LibXML::Nodish does XML::LibXML::C14N {
+
+        # cw: ?????
+        #PmmFixOwner( SvPROXYNODE(RETVAL), NULL );
+        return $retVal;
+    }
+
+    method setAttributeNodeNS(xmlAttr $an!) {
+        sub xmlReconciliateNs(xmlDoc, xmlNode) returns int32 is native('xml2') { * }
+
+        if !$an.defined {
+            die "lost attribute node";
+            # cw: For .resume in CATCH{}
+            return;
+        }
+
+        return unless $an.type != XML_ATTRIBUTE_NODE;
+
+        domImportNode(self.doc, $an, 1, 1) if $an.doc !=:= self.doc;
+        my xmlNs $ns = $an.ns;
+        my $ret = xmlHasNsProp(
+            self, 
+            $ns.defined ?? $ns.uri !! Str,
+            $an.localname
+        );
+
+        if $ret.defined && $ret.type == XML_ATTRIBUTE_NODE {
+            return if $ret =:= $an;
+            xmlReplaceNode($ret, $an);
+        } else {
+            xmlAddChild(
+                nativecast(xmlNodePtr, self), 
+                nativecast(xmlNodePtr, $an)
+            );
+            xmlReconciliateNs(self.doc, self);
+        }
+
+        # cw: ??? XS
+        #if ( attr->_private != NULL ) {
+        #    PmmFixOwner( SvPROXYNODE(attr_node), PmmPROXYNODE(self) );
+        #}
+
+        return $ret;
+    }
+
+    method removeAttributeNode(xmlAttr $an!) {
+        if !$an.defined {
+            die "lost attribute node";
+            # cw: For .resume in CATCH{}
+            return;
+        }
+
+        return unless 
+            $an.type == XML_ATTRIBUTE_NODE
+            &&
+            $an.parent !=:= self;
+
+        my $an_p = nativecast(xmlNodePtr, $an);
+        xmlUnlinkNode($an_p);
+
+        # cw: ????
+        # $ret = PmmNodeToSv($ret_p, NUL)
+        # PmmFixOwner( SvPROXYNODE($ret), NULL)
+
+        return $an;
+    }
+
+    method removeAttributeNS($_nsUri, $_name) {
+        sub xmlFreeProp(xmlAttrPtr) is native('xml2') { * }
+
+        my $nsUri = $_nsUri.defined ?? $_nsUri.subst(/\s/, '') !! Str;
+        my $name = $_name.defined ?? $_name.subst(/\s/, '') !! Str;
+
+        my xmlAttr $xattr = xmlHasNsProp(
+            self.getNode(), $name, $nsUri
+        );
+        xmlUnlinkNode(nativecast(xmlNodePtr, $xattr)) 
+            if $xattr.defined && $xattr.type == XML_ATTRIBUTE_NODE;
+
+        if ($xattr.defined && $xattr._private.defined) {
+            # cw: ???? XS code
+            # PmmFixOwner((ProxyNodePtr)xattr->_private, NULL);
+        } else {
+            xmlFreeProp(nativecast(xmlAttrPtr, $xattr));
+        }
+    }
+
+    method setNamespace($_uri, $_prefix = Nil, $flag = 1) {
+        my $ret = 0;
+        my $ns;
+
+        if (! $_uri.defined && ! $_prefix.defined) {
+            $ns = xmlSearchNs(self.doc, self.getNode(), Str);
+            if ($ns.defined && $ns.uri.defined && $ns.uri.chars) {
+                $ret = 0;
+            } elsif $flag {
+                xmlSetNs(self.getNode(), xmlNs);
+                $ret = 1;
+            } else {
+                $ret = 0;
+            }
+        } elsif $flag {
```

Comment:
This entire function got a WTF from me. I basically ported it straight from the XS code and then tried to simplify it. This is probably one of the situations where we diverge from what is in XML::LibXML.

I will look into this one again and see if it can be rewritten concisely.